File: zms-be/animal-service/src/main/java/it/zoo/animal/domain/model/Animal.java
package it.zoo.animal.domain.model;

import java.time.LocalDate;
import java.util.UUID;

public class Animal {

    private UUID id;
    private String name;
    private String species;
    private boolean dangerous;
    private Habitat habitat;
    private UUID enclosureId;
    private LocalDate arrivalDate;
    private AnimalStatus status;

    public Animal() {}

    public Animal(UUID id, String name, String species, boolean dangerous,
                  Habitat habitat, UUID enclosureId, LocalDate arrivalDate,
                  AnimalStatus status) {
        this.id = id;
        this.name = name;
        this.species = species;
        this.dangerous = dangerous;
        this.habitat = habitat;
        this.enclosureId = enclosureId;
        this.arrivalDate = arrivalDate;
        this.status = status;
    }

    public UUID getId() { return id; }
    public String getName() { return name; }
    public String getSpecies() { return species; }
    public boolean isDangerous() { return dangerous; }
    public Habitat getHabitat() { return habitat; }
    public UUID getEnclosureId() { return enclosureId; }
    public LocalDate getArrivalDate() { return arrivalDate; }
    public AnimalStatus getStatus() { return status; }

    public void setId(UUID id) { this.id = id; }
    public void setName(String name) { this.name = name; }
    public void setSpecies(String species) { this.species = species; }
    public void setDangerous(boolean dangerous) { this.dangerous = dangerous; }
    public void setHabitat(Habitat habitat) { this.habitat = habitat; }
    public void setEnclosureId(UUID enclosureId) { this.enclosureId = enclosureId; }
    public void setArrivalDate(LocalDate arrivalDate) { this.arrivalDate = arrivalDate; }
    public void setStatus(AnimalStatus status) { this.status = status; }

    public boolean canTransitionTo(AnimalStatus target) {
        if (this.status == AnimalStatus.DECEASED) {
            return false;
        }
        return this.status != target;
    }
}

File: zms-fe/src/app/core/models/animal.ts
/** Mirrors `it.zoo.animal.domain.enums.AnimalStatus`. */
export type AnimalStatus = 'HEALTHY' | 'UNDER_OBSERVATION' | 'IN_TREATMENT' | 'DECEASED';

/** Mirrors `it.zoo.animal.domain.enums.Habitat`. */
export type Habitat = 'TERRESTRIAL' | 'AQUATIC' | 'AMPHIBIOUS';

/** Keycloak realm roles enforced by `AnimalResource`. */
export type ZooRole = 'zoo-keeper' | 'zoo-vet' | 'zoo-admin';

/** Mirrors `AnimalResponse` from animal-service. */
export interface Animal {
  readonly id: string;
  readonly name: string;
  readonly species: string;
  readonly dangerous: boolean;
  readonly habitat: Habitat;
  readonly enclosureId: string;
  readonly arrivalDate: string;
  readonly status: AnimalStatus;
  readonly createdBy: string | null;
  readonly updatedBy: string | null;
}

/**
 * Enclosures exist only as UUIDs in the backend. The name and habitat live
 * in the frontend directory until an enclosure resource exists.
 */
export interface Enclosure {
  readonly id: string;
  readonly name: string;
  readonly habitat: Habitat;
}

export const STATUS_ORDER: readonly AnimalStatus[] = [
  'HEALTHY',
  'UNDER_OBSERVATION',
  'IN_TREATMENT',
  'DECEASED',
];

/** Same rule as `Animal.canTransitionTo` in the domain model. */
export function canTransitionTo(animal: Animal, target: AnimalStatus): boolean {
  return animal.status !== 'DECEASED' && animal.status !== target;
}

/** Same rule as `Animal.canBeTransferred` in the domain model. */
export function canBeTransferred(animal: Animal): boolean {
  return animal.status !== 'DECEASED';
}

/** Short, human-readable tag code derived from the animal's UUID. */
export function tagCode(animal: Pick<Animal, 'id'>): string {
  return animal.id.slice(0, 4).toUpperCase();
}
